### PR TITLE
Fix stereo -> mono downmixing.

### LINKED ISSUE #240: (Debug builds) Reproducible crash at buffer_manager.rs:59 with S16NE stereo input device downmixing to mono
If you have a stereo input device that uses S16 data, and request a mono stream, cubeb_coreaudio_rs will reproducibly crash in debug builds if the sound level is high enough that the addition on [buffer_manager.rs:59](https://github.com/mozilla/cubeb-coreaudio-rs/blob/f70ef45e66f51722f56adac79e6710ae079f7bd4/src/backend/buffer_manager.rs#L59) overflows.

I believe the addition there should be a `saturating_add`, instead. (or ideally an average?)

For example, using https://github.com/mutexlox-signal/test-coreaudio-rs/ hooked up to a [BlackHole 2-channel input](https://github.com/ExistentialAudio/BlackHole) (which is fed from a stereo youtube video):

```
thread '<unnamed>' panicked at src/backend/buffer_manager.rs:59:31:
attempt to add with overflow
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at library/core/src/panicking.rs:221:5:
panic in a function that cannot unwind
stack backtrace:
   0:        0x104858654 - std::backtrace_rs::backtrace::libunwind::trace::hbebc8679d47bdc2c
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:        0x104858654 - std::backtrace_rs::backtrace::trace_unsynchronized::h3a2e9637943241aa
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:        0x104858654 - std::sys::backtrace::_print_fmt::he430849680584674
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/sys/backtrace.rs:65:5
   3:        0x104858654 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h243268f17d714c7f
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/sys/backtrace.rs:40:26
   4:        0x1048706b4 - core::fmt::rt::Argument::fmt::h0d339881c25f3c31
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/fmt/rt.rs:173:76
   5:        0x1048706b4 - core::fmt::write::hb3cfb8a30e72d7ff
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/fmt/mod.rs:1182:21
   6:        0x104856760 - std::io::Write::write_fmt::hfb2314975de9ecf1
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/io/mod.rs:1827:15
   7:        0x104859730 - std::sys::backtrace::BacktraceLock::print::he14461129ccbfef5
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/sys/backtrace.rs:43:9
   8:        0x104859730 - std::panicking::default_hook::{{closure}}::h14c7718ccf39d316
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:269:22
   9:        0x104859354 - std::panicking::default_hook::hc62e60da3be2f352
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:296:9
  10:        0x10485a1f4 - std::panicking::rust_panic_with_hook::h09e8a656f11e82b2
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:800:13
  11:        0x104859b1c - std::panicking::begin_panic_handler::{{closure}}::h1230eb3cc91b241c
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:667:13
  12:        0x104858ae0 - std::sys::backtrace::__rust_end_short_backtrace::hc3491307aceda2c2
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/sys/backtrace.rs:168:18
  13:        0x10485980c - rust_begin_unwind
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:665:5
  14:        0x104877e60 - core::panicking::panic_nounwind_fmt::runtime::h5290ab2b4897aadc
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:112:18
  15:        0x104877e60 - core::panicking::panic_nounwind_fmt::h91ee161184879b56
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:122:5
  16:        0x104877ed8 - core::panicking::panic_nounwind::heab7ebe7a6cd845c
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:221:5
  17:        0x104878008 - core::panicking::panic_cannot_unwind::hedc43d82620205bf
                               at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:309:5
  18:        0x1047f3c50 - cubeb_coreaudio::backend::audiounit_input_callback::h98caba4ee54ce43a
                               at /Users/miriam/cubeb-rs/cubeb-sys/libcubeb/src/cubeb-coreaudio-rs/src/backend/mod.rs:481:1
  19:        0x106dae5b0 - _AUGenericOutputFactory
  20:        0x1878143d4 - <unknown>
  21:        0x187811a8c - <unknown>
  22:        0x187997850 - <unknown>
  23:        0x184fa1f94 - __pthread_joiner_wake
thread caused non-unwinding panic. aborting.
```

### PR DESCRIPTION
**Pull request by mutexlox-signal**
Rather than summing the two channels, average them. This prevents integer overflow and clipping or needlessly increasing volume.

This will fix #240.

### CODE REVIEW

File: src/backend/buffer_manager.rs [hunk 1]
```diff
@@ -36,17 +37,34 @@ fn drop_first_n_channels_in_place<T: Copy>(
...
-fn remix_or_drop_channels<T: Copy + std::ops::Add<Output = T>>(
+fn remix_or_drop_channels<T: DataType<T>>(
     input_channels: usize,
     output_channels: usize,
     data: &mut [T],
     frame_count: usize,
-) -> usize {
+) -> usize
+where
+    f32: AsPrimitive<T>,
```
Comment (mutexlox-signal) [inline src/backend/buffer_manager.rs]: I'm not thrilled that I need to repeat this bound here (and below), and I don't quite understand why having it on the trait and its impl is insufficient; thoughts appreciated.

**Comment by mutexlox-signal** [comment]
cc @padenot 

not sure what the failures are, but they at least seem unrelated? i think?

File: src/backend/buffer_manager.rs (hunk 1 above)
Comment (mutexlox-signal) [reply src/backend/buffer_manager.rs]: Resolved.

**Review by padenot — APPROVED** [review]

**Comment by mutexlox-signal** [comment]
I'm curious whether there's e.g. a breakage with a new version of macOS 14, or a new rustc or something, that is causing the breakage on macOS 14? I looked briefly at the failing tests, and I don't see an obvious connection. I wonder whether running the tests at HEAD on trailblazer would show the same issue.

**Comment by padenot** [comment]
Historically there's been breakage related to upgrades. https://github.com/mozilla/cubeb-coreaudio-rs/actions/runs/11283677817/job/31383525624 shows that you're right.

**Comment by mutexlox-signal** [comment]
Thanks!

**Comment by mutexlox-signal** [comment]
Would you expect that this breakage is scoped just to tests, or do you think it will affect real-world usage as well?

**Comment by Pehrsons** [comment]
> Would you expect that this breakage is scoped just to tests, or do you think it will affect real-world usage as well?

This looks like an issue with the test-only device plugger. Or an issue/difference with macOS 14. Also it's running in a VM and sometimes they seem to get into a weird state because tests that normally don't fail suddenly fail at a high rate.

**Comment by mutexlox-signal** [comment]
Thanks!

**Pull request merged**